File: hydra-java/src/main/java/hydra/dsl/Expect.java
package hydra.dsl;

import hydra.Reduction;
import hydra.compute.Flow;
import hydra.core.Field;
import hydra.core.FloatValue;
import hydra.core.IntegerValue;
import hydra.core.Literal;
import hydra.core.Name;
import hydra.core.Term;
import hydra.core.Type;
import hydra.graph.Graph;
import hydra.tools.PrettyPrinter;
import hydra.util.Opt;
import hydra.util.Tuple;
import java.math.BigInteger;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;

import static hydra.dsl.Flows.bind;
import static hydra.dsl.Flows.fail;
import static hydra.dsl.Flows.mapM;
import static hydra.dsl.Flows.pure;
import static hydra.dsl.Flows.unexpected;


/**
 * Functions for decoding of native Java values from Hydra terms.
 */
public class Expect {
    private Expect() {
    }

    /**
     * Decode a bigfloat value.
     */
    public static <S> Flow<S, Double> bigfloat(final Term term) {
        return bind(float_(term), floatValue -> floatValue.accept(new FloatValue.PartialVisitor<Flow<S, Double>>() {
            @Override
            public Flow<S, Double> otherwise(FloatValue instance) {
                return wrongType("bigfloat", term);
            }

            @Override
            public Flow<S, Double> visit(FloatValue.Bigfloat instance) {
                return pure(instance.value);
            }
        }));
    }

    /**
     * Decode a bigint value.
     */
    public static <S> Flow<S, BigInteger> bigint(final Term term) {
        return bind(integer(term),
                integerValue -> integerValue.accept(new IntegerValue.PartialVisitor<Flow<S, BigInteger>>() {
                    @Override
                    public Flow<S, BigInteger> otherwise(IntegerValue instance) {
                        return wrongType("bigint", term);
                    }

                    @Override
                    public Flow<S, BigInteger> visit(IntegerValue.Bigint instance) {
                        return pure(instance.value);
                    }
                }));
    }

    /**
     * Decode a binary value.
     */
    public static <S> Flow<S, String> binary(final Term term) {
        return bind(literal(term), literal -> literal.accept(new Literal.PartialVisitor<Flow<S, String>>() {
            @Override
            public Flow<S, String> otherwise(Literal instance) {
                return wrongType("binary", term);
            }

            @Override
            public Flow<S, String> visit(Literal.Binary instance) {
                return pure(instance.value);
            }
        }));
    }

    /**
     * Decode a boolean value.
     */
    public static <S> Flow<S, Boolean> boolean_(final Term term) {
        return bind(literal(term), literal -> literal.accept(new Literal.PartialVisitor<Flow<S, Boolean>>() {
            @Override
            public Flow<S, Boolean> otherwise(Literal instance) {
                return wrongType("boolean", term);
            }

            @Override
            public Flow<S, Boolean> visit(Literal.Boolean_ instance) {
                return pure(instance.value);
            }
        }));
    }

    /**
     * Retrieve and decode a field from a map of field names to terms.
     */
    public static <S, X> Flow<S, X> field(final Name fname,
                                          final Function<Term, Flow<S, X>> accessor,
                                          final Map<Name, Term> fields) {
        Term term = fields.get(fname);
        if (term == null) {
            return Flows.fail("field " + fname + " not found");
        } else {
            return accessor.apply(term);
        }
    }

    /**
     * Decode a floating point value.
     */
    public static <S> Flow<S, FloatValue> float_(final Term term) {
        return bind(literal(term), literal -> literal.accept(new Literal.PartialVisitor<Flow<S, FloatValue>>() {
            @Override
            public Flow<S, FloatValue> otherwise(Literal instance) {
                return wrongType("float", term);
            }

            @Override
            public Flow<S, FloatValue> visit(Literal.Float_ instance) {
                return pure(instance.value);
            }
        }));
    }

    /**
     * Decode a float32 value.
     */
    public static <S> Flow<S, Float> float32(final Term term) {
        return bind(float_(term), floatValue -> floatValue.accept(new FloatValue.PartialVisitor<Flow<S, Float>>() {
            @Override
            public Flow<S, Float> otherwise(FloatValue instance) {
                return wrongType("float32", term);
            }

            @Override
            public Flow<S, Float> visit(FloatValue.Float32 instance) {
                return pure(instance.value);
            }
        }));
    }

    /**
     * Decode a float64 value.
     */
    public static <S> Flow<S, Double> float64(final Term term) {
        return bind(float_(term), floatValue -> floatValue.accept(new FloatValue.PartialVisitor<Flow<S, Double>>() {
            @Override
            public Flow<S, Double> otherwise(FloatValue instance) {
                return wrongType("float64", term);
            }

            @Override
            public Flow<S, Double> visit(FloatValue.Float64 instance) {
                return pure(instance.value);
            }
        }));
    }

    /**
     * Decode a function.
     */
    public static <X, Y> Function<X, Flow<Graph, Y>> function(
            final Function<X, Term> fin,
            final Function<Term, Flow<Graph, Y>> fout,
            final Term func) {
        return x -> bind(Reduction.reduce(false, Terms.apply(func, fin.apply(x))), fout);
    }

    /**
     * Decode an int8 value.
     */
    public static <S> Flow<S, Byte> int8(final Term term) {
        return bind(integer(term),
                integerValue -> integerValue.accept(new IntegerValue.PartialVisitor<Flow<S, Byte>>() {
                    @Override
                    public Flow<S, Byte> otherwise(IntegerValue instance) {
                        return wrongType("int8", term);
                    }

                    @Override
                    public Flow<S, Byte> visit(IntegerValue.Int8 instance) {
                        return pure(instance.value);
                    }
                }));
    }

    /**
     * Decode an int16 value.
     */
    public static <S> Flow<S, Short> int16(final Term term) {
        return bind(integer(term),
                integerValue -> integerValue.accept(new IntegerValue.PartialVisitor<Flow<S, Short>>() {
                    @Override
                    public Flow<S, Short> otherwise(IntegerValue instance) {
                        return wrongType("int16", term);
                    }

                    @Override
                    public Flow<S, Short> visit(IntegerValue.Int16 instance) {
                        return pure(instance.value);
                    }
                }));
    }

    /**
     * Decode an int32 value.
     */
    public static <S> Flow<S, Integer> int32(final Term term) {
        return bind(integer(term),
                integerValue -> integerValue.accept(new IntegerValue.PartialVisitor<Flow<S, Integer>>() {
                    @Override
                    public Flow<S, Integer> otherwise(IntegerValue instance) {
                        return wrongType("int32", term);
                    }

                    @Override
                    public Flow<S, Integer> visit(IntegerValue.Int32 instance) {
                        return pure(instance.value);
                    }
                }));
    }

    /**
     * Decode an int64 value.
     */
    public static <S> Flow<S, Long> int64(final Term term) {
        return bind(integer(term),
                integerValue -> integerValue.accept(new IntegerValue.PartialVisitor<Flow<S, Long>>() {
                    @Override
                    public Flow<S, Long> otherwise(IntegerValue instance) {
                        return wrongType("int64", term);
                    }

                    @Override
                    public Flow<S, Long> visit(IntegerValue.Int64 instance) {
                        return pure(instance.value);
                    }
                }));
    }

    /**
     * Decode an integer value.
     */
    public static <S> Flow<S, IntegerValue> integer(final Term term) {
        return bind(literal(term), literal -> literal.accept(new Literal.PartialVisitor<Flow<S, IntegerValue>>() {
            @Override
            public Flow<S, IntegerValue> otherwise(Literal instance) {
                return wrongType("float", term);
            }

            @Override
            public Flow<S, IntegerValue> visit(Literal.Integer_ instance) {
                return pure(instance.value);
            }
        }));
    }

    /**
     * Decode a list of values.
     */
    public static <S, X> Flow<S, List<X>> list(final Function<Term, Flow<S, X>> elems, final Term term) {
        return term.accept(new Term.PartialVisitor<Flow<S, List<X>>>() {
            @Override
            public Flow<S, List<X>> otherwise(Term instance) {
                return wrongType("list", term);
            }

            @Override
            public Flow<S, List<X>> visit(Term.List instance) {
                return mapM(instance.value, elems);
            }
        });
    }

    /**
     * Decode a literal value.
     */
    public static <S> Flow<S, Literal> literal(final Term term) {
        return term.accept(new Term.PartialVisitor<Flow<S, Literal>>() {
            @Override
            public Flow<S, Literal> otherwise(Term instance) {
                return wrongType("literal", term);
            }

            @Override
            public Flow<S, Literal> visit(Term.Literal instance) {
                return pure(instance.value);
            }
        });
    }

    /**
     * Decode a map of keys to values.
     */
    public static <S, K, V> Flow<S, Map<K, V>> map(
            final Function<Term, Flow<S, K>> keys,
            final Function<Term, Flow<S, V>> values,
            final Term term) {
        return term.accept(new Term.PartialVisitor<Flow<S, Map<K, V>>>() {
            @Override
            public Flow<S, Map<K, V>> otherwise(Term instance) {
                return wrongType("map", term);
            }

            @Override
            public Flow<S, Map<K, V>> visit(Term.Map instance) {
                return Flows.map(Flows.mapM(new ArrayList<>(instance.value.entrySet()),
                                entry -> Flows.map2(
                                        keys.apply(entry.getKey()),
                                        values.apply(entry.getValue()),
                                        AbstractMap.SimpleEntry::new)),
                        (Function<List<Map.Entry<K, V>>, Map<K, V>>) entries -> {
                            Map<K, V> result = new HashMap<>();
                            for (Map.Entry<K, V> e : entries) {
                                result.put(e.getKey(), e.getValue());
                            }
                            return result;
                        });
            }
        });
    }

    /**
     * Decode an optional value.
     */
    public static <S, X> Flow<S, Opt<X>> optional(final Function<Term, Flow<S, X>> elems,
                                                  final Term term) {
        return term.accept(new Term.PartialVisitor<Flow<S, Opt<X>>>() {
            @Override
            public Flow<S, Opt<X>> otherwise(Term instance) {
                return wrongType("optional", term);
            }

            @Override
            public Flow<S, Opt<X>> visit(Term.Optional instance) {
                return instance.value.isPresent() ? Flows.map(elems.apply(instance.value.get()), Opt::of)
                        : pure(Opt.empty());
            }
        });
    }

    /**
     * Decode a pair of values.
     */
    public static <S, T1, T2> Flow<S, Tuple.Tuple2<T1, T2>> pair(
            final Function<Term, Flow<S, T1>> first,
            final Function<Term, Flow<S, T2>> second,
            final Term term) {
        return term.accept(new Term.PartialVisitor<Flow<S, Tuple.Tuple2<T1, T2>>>() {
            @Override
            public Flow<S, Tuple.Tuple2<T1, T2>> otherwise(Term instance) {
                return wrongType("tuple", term);
            }

            @Override
            public Flow<S, Tuple.Tuple2<T1, T2>> visit(Term.Product instance) {
                List<Term> values = instance.value;
                if (values.size() != 2) {
                    return fail("Expected a tuple of size 2, but found " + values.size());
                }
                return Flows.map2(first.apply(values.get(0)), second.apply(values.get(1)), Tuple.Tuple2::new);
            }
        });
    }

    /**
     * Decode a record.
     */
    public static <S> Flow<S, List<Field>> record(final Name tname, final Term term) {
        return term.accept(new Term.PartialVisitor<Flow<S, List<Field>>>() {
            @Override
            public Flow<S, List<Field>> otherwise(Term instance) {
                return wrongType("record", term);
            }

            @Override
            public Flow<S, List<Field>> visit(Term.Record instance) {
                if (instance.value.typeName.equals(tname)) {
                    return pure(instance.value.fields);
                } else {
                    return fail("Expected a record of type " + tname + ", but found " + instance.value.typeName);
                }
            }
        });
    }

    /**
     * Decode a record as a map from field names to terms.
     */
    public static <S> Flow<S, Map<Name, Term>> recordAsMap(final Name tname, final Term term) {
        return Flows.map(record(tname, term), fields -> {
            Map<Name, Term> result = new HashMap<>();
            for (Field f : fields) {
                result.put(f.name, f.term);
            }
            return result;
        });
    }

    /**
     * Decode a set of values.
     */
    public static <S, X> Flow<S, Set<X>> set(final Function<Term, Flow<S, X>> elems, final Term term) {
        return term.accept(new Term.PartialVisitor<Flow<S, Set<X>>>() {
            @Override
            public Flow<S, Set<X>> otherwise(Term instance) {
                return wrongType("set", term);
            }

            @Override
            public Flow<S, Set<X>> visit(Term.Set instance) {
                return mapM(instance.value, elems);
            }
        });
    }

    /**
     * Decode a string value.
     */
    public static <S> Flow<S, String> string(final Term term) {
        return bind(literal(term), literal -> literal.accept(new Literal.PartialVisitor<Flow<S, String>>() {
            @Override
            public Flow<S, String> otherwise(Literal instance) {
                return wrongType("string", term);
            }

            @Override
            public Flow<S, String> visit(Literal.String_ instance) {
                return pure(instance.value);
            }
        }));
    }

    /**
     * Decode a term.
     */
    public static <S> Flow<S, Term> term(final Term term) {
        return pure(term);
    }

    /**
     * Decode a type.
     */
    public static <S> Flow<S, Type> type(final Term term) {
        return fail("Core decoding not yet implemented");
    }

    /**
     * Decode a uint8 value.
     */
    public static <S> Flow<S, Character> uint8(final Term term) {
        return bind(integer(term),
                integerValue -> integerValue.accept(new IntegerValue.PartialVisitor<Flow<S, Character>>() {
                    @Override
                    public Flow<S, Character> otherwise(IntegerValue instance) {
                        return wrongType("uint8", term);
                    }

                    @Override
                    public Flow<S, Character> visit(IntegerValue.Uint8 instance) {
                        return pure(instance.value);
                    }
                }));
    }

    /**
     * Decode a uint16 value.
     */
    public static <S> Flow<S, Character> uint16(final Term term) {
        return bind(integer(term),
                integerValue -> integerValue.accept(new IntegerValue.PartialVisitor<Flow<S, Character>>() {
                    @Override
                    public Flow<S, Character> otherwise(IntegerValue instance) {
                        return wrongType("uint16", term);
                    }

                    @Override
                    public Flow<S, Character> visit(IntegerValue.Uint16 instance) {
                        return pure(instance.value);
                    }
                }));
    }

    /**
     * Decode a uint32 value.
     */
    public static <S> Flow<S, Long> uint32(final Term term) {
        return bind(integer(term),
                integerValue -> integerValue.accept(new IntegerValue.PartialVisitor<Flow<S, Long>>() {
                    @Override
                    public Flow<S, Long> otherwise(IntegerValue instance) {
                        return wrongType("uint32", term);
                    }

                    @Override
                    public Flow<S, Long> visit(IntegerValue.Uint32 instance) {
                        return pure(instance.value);
                    }
                }));
    }

    /**
     * Decode a uint64 value.
     */
    public static <S> Flow<S, BigInteger> uint64(final Term term) {
        return bind(integer(term),
                integerValue -> integerValue.accept(new IntegerValue.PartialVisitor<Flow<S, BigInteger>>() {
                    @Override
                    public Flow<S, BigInteger> otherwise(IntegerValue instance) {
                        return wrongType("uint64", term);
                    }

                    @Override
                    public Flow<S, BigInteger> visit(IntegerValue.Uint64 instance) {
                        return pure(instance.value);
                    }
                }));
    }

    private static <S, X> Flow<S, X> wrongType(String category, Term term) {
        return unexpected(category, PrettyPrinter.printTerm(term));
    }
}

File: hydra-haskell/src/main/haskell/Hydra/Dsl/Expect.hs
-- | A DSL for decoding and validating Hydra terms at runtime.
-- This module provides functions to extract typed values from Hydra terms
-- with appropriate error handling.
module Hydra.Dsl.Expect where

import Hydra.Compute
import Hydra.Core
import Hydra.Graph
import Hydra.Strip
import Hydra.Rewriting
import Hydra.Errors
import Hydra.Lexical
import qualified Hydra.Lib.Flows as Flows

import Prelude hiding (map)
import qualified Data.List as L
import qualified Data.Map as M
import qualified Data.Set as S
import qualified Data.Maybe as Y
import qualified Control.Monad as CM
import Data.Int

-- * Validation

-- | Ensure a function has the expected number of arguments
-- Example: nArgs _add 2 args
nArgs :: Name -> Int -> [Term] -> Flow s ()
nArgs name n args = if L.length args /= n
  then unexpected (show n ++ " arguments to primitive " ++ show (unName name)) $ show (L.length args)
  else pure ()

-- * Literal values

-- | Extract a binary data value from a term
-- Example: binary term
binary :: Term -> Flow Graph String
binary t = literal t >>= binaryLiteral

-- | Extract a binary literal from a Literal value
-- Example: literal term >>= binaryLiteral
binaryLiteral :: Literal -> Flow Graph String
binaryLiteral v = case v of
  LiteralBinary b -> pure b
  _ -> unexpected "binary" $ show v

-- | Extract a boolean value from a term
-- Example: boolean term
boolean :: Term -> Flow Graph Bool
boolean t = literal t >>= booleanLiteral

-- | Extract a boolean literal from a Literal value
-- Example: literal term >>= booleanLiteral
booleanLiteral :: Literal -> Flow Graph Bool
booleanLiteral v = case v of
  LiteralBoolean b -> pure b
  _ -> unexpected "boolean" $ show v

-- | Extract a literal value from a term
-- Example: literal term >>= stringLiteral
literal :: Term -> Flow Graph Literal
literal term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermLiteral lit -> pure lit
    _ -> unexpected "literal" $ show term

-- | Extract a string value from a term
-- Example: string term
string :: Term -> Flow Graph String
string t = literal t >>= stringLiteral

-- | Extract a string literal from a Literal value
-- Example: literal term >>= stringLiteral
stringLiteral :: Literal -> Flow Graph String
stringLiteral v = case v of
  LiteralString s -> pure s
  _ -> unexpected "string" $ show v

-- ** Integers

-- | Extract an arbitrary-precision integer value from a term
-- Example: bigint term
bigint :: Term -> Flow Graph Integer
bigint t = literal t >>= integerLiteral >>= bigintValue

-- | Extract a bigint value from an IntegerValue
-- Example: integerLiteral lit >>= bigintValue
bigintValue :: IntegerValue -> Flow Graph Integer
bigintValue v = case v of
  IntegerValueBigint i -> pure i
  _ -> unexpected "bigint" $ show v

-- | Extract an integer literal from a Literal value
-- Example: literal term >>= integerLiteral
integerLiteral :: Literal -> Flow Graph IntegerValue
integerLiteral lit = case lit of
  LiteralInteger v -> pure v
  _ -> unexpected "integer value" $ show lit

-- | Extract an 8-bit signed integer value from a term
-- Example: int8 term
int8 :: Term -> Flow Graph Int8
int8 t = literal t >>= integerLiteral >>= int8Value

-- | Extract an int8 value from an IntegerValue
-- Example: integerLiteral lit >>= int8Value
int8Value :: IntegerValue -> Flow Graph Int8
int8Value v = case v of
  IntegerValueInt8 i -> pure i
  _ -> unexpected "int8" $ show v

-- | Extract a 16-bit signed integer value from a term
-- Example: int16 term
int16 :: Term -> Flow Graph Int16
int16 t = literal t >>= integerLiteral >>= int16Value

-- | Extract an int16 value from an IntegerValue
-- Example: integerLiteral lit >>= int16Value
int16Value :: IntegerValue -> Flow Graph Int16
int16Value v = case v of
  IntegerValueInt16 i -> pure i
  _ -> unexpected "int16" $ show v

-- | Extract a 32-bit signed integer value from a term
-- Example: int32 term
int32 :: Term -> Flow Graph Int
int32 t = literal t >>= integerLiteral >>= int32Value

-- | Extract an int32 value from an IntegerValue
-- Example: integerLiteral lit >>= int32Value
int32Value :: IntegerValue -> Flow Graph Int
int32Value v = case v of
  IntegerValueInt32 i -> pure i
  _ -> unexpected "int32" $ show v

-- | Extract a 64-bit signed integer value from a term
-- Example: int64 term
int64 :: Term -> Flow Graph Int64
int64 t = literal t >>= integerLiteral >>= int64Value

-- | Extract an int64 value from an IntegerValue
-- Example: integerLiteral lit >>= int64Value
int64Value :: IntegerValue -> Flow Graph Int64
int64Value v = case v of
  IntegerValueInt64 i -> pure i
  _ -> unexpected "int64" $ show v

-- | Extract an 8-bit unsigned integer value from a term
-- Example: uint8 term
uint8 :: Term -> Flow Graph Int16
uint8 t = literal t >>= integerLiteral >>= uint8Value

-- | Extract a uint8 value from an IntegerValue
-- Example: integerLiteral lit >>= uint8Value
uint8Value :: IntegerValue -> Flow Graph Int16
uint8Value v = case v of
  IntegerValueUint8 i -> pure i
  _ -> unexpected "uint8" $ show v

-- | Extract a 16-bit unsigned integer value from a term
-- Example: uint16 term
uint16 :: Term -> Flow Graph Int
uint16 t = literal t >>= integerLiteral >>= uint16Value

-- | Extract a uint16 value from an IntegerValue
-- Example: integerLiteral lit >>= uint16Value
uint16Value :: IntegerValue -> Flow Graph Int
uint16Value v = case v of
  IntegerValueUint16 i -> pure i
  _ -> unexpected "uint16" $ show v

-- | Extract a 32-bit unsigned integer value from a term
-- Example: uint32 term
uint32 :: Term -> Flow Graph Int64
uint32 t = literal t >>= integerLiteral >>= uint32Value

-- | Extract a uint32 value from an IntegerValue
-- Example: integerLiteral lit >>= uint32Value
uint32Value :: IntegerValue -> Flow Graph Int64
uint32Value v = case v of
  IntegerValueUint32 i -> pure i
  _ -> unexpected "uint32" $ show v

-- | Extract a 64-bit unsigned integer value from a term
-- Example: uint64 term
uint64 :: Term -> Flow Graph Integer
uint64 t = literal t >>= integerLiteral >>= uint64Value

-- | Extract a uint64 value from an IntegerValue
-- Example: integerLiteral lit >>= uint64Value
uint64Value :: IntegerValue -> Flow Graph Integer
uint64Value v = case v of
  IntegerValueUint64 i -> pure i
  _ -> unexpected "uint64" $ show v

-- ** Floating-point values

-- | Extract an arbitrary-precision floating-point value from a term
-- Example: bigfloat term
bigfloat :: Term -> Flow Graph Double
bigfloat t = literal t >>= floatLiteral >>= bigfloatValue

-- | Extract a bigfloat value from a FloatValue
-- Example: floatLiteral lit >>= bigfloatValue
bigfloatValue :: FloatValue -> Flow Graph Double
bigfloatValue v = case v of
  FloatValueBigfloat f -> pure f
  _ -> unexpected "bigfloat" $ show v

-- | Extract a floating-point literal from a Literal value
-- Example: literal term >>= floatLiteral
floatLiteral :: Literal -> Flow Graph FloatValue
floatLiteral lit = case lit of
  LiteralFloat v -> pure v
  _ -> unexpected "floating-point value" $ show lit

-- | Extract a 32-bit floating-point value from a term
-- Example: float32 term
float32 :: Term -> Flow Graph Float
float32 t = literal t >>= floatLiteral >>= float32Value

-- | Extract a float32 value from a FloatValue
-- Example: floatLiteral lit >>= float32Value
float32Value :: FloatValue -> Flow Graph Float
float32Value v = case v of
  FloatValueFloat32 f -> pure f
  _ -> unexpected "float32" $ show v

-- | Extract a 64-bit floating-point value from a term
-- Example: float64 term
float64 :: Term -> Flow Graph Double
float64 t = literal t >>= floatLiteral >>= float64Value

-- | Extract a float64 value from a FloatValue
-- Example: floatLiteral lit >>= float64Value
float64Value :: FloatValue -> Flow Graph Double
float64Value v = case v of
  FloatValueFloat64 f -> pure f
  _ -> unexpected "float64" $ show v

-- * Collections

-- | Extract a list of values from a term, mapping a function over each element
-- Example: list string term
list :: (Term -> Flow Graph x) -> Term -> Flow Graph [x]
list f term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermList l -> CM.mapM f l
    _ -> unexpected "list" $ show term

-- | Extract the first element of a list term
-- Example: listHead term
listHead :: Term -> Flow Graph Term
listHead term = do
  l <- list pure term
  if L.null l
    then fail "empty list"
    else pure $ L.head l

-- | Extract the element type from a list type
-- Example: listType typ
listType :: Type -> Flow s Type
listType typ = case stripType typ of
  TypeList t -> pure t
  _ -> unexpected "list type" $ show typ

-- | Extract a map of key-value pairs from a term, mapping functions over each key and value
-- Example: map string int32 term
map :: Ord k => (Term -> Flow Graph k) -> (Term -> Flow Graph v) -> Term -> Flow Graph (M.Map k v)
map fk fv term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermMap m -> M.fromList <$> CM.mapM pair (M.toList m)
      where
        pair (kterm, vterm) = do
          kval <- fk kterm
          vval <- fv vterm
          return (kval, vval)
    _ -> unexpected "map" $ show term

-- | Extract the key and value types from a map type
-- Example: mapType typ
mapType :: Type -> Flow s MapType
mapType typ = case stripType typ of
  TypeMap mt -> pure mt
  _ -> unexpected "map type" $ show typ

-- | Extract an optional value from a term, applying a function to the value if present
-- Example: optional string term
optional :: (Term -> Flow Graph x) -> Term -> Flow Graph (Y.Maybe x)
optional f term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermOptional mt -> case mt of
      Nothing -> pure Nothing
      Just t -> Just <$> f t
    _ -> unexpected "optional value" $ show term

-- | Extract the base type from an optional type
-- Example: optionalType typ
optionalType :: Type -> Flow s Type
optionalType typ = case stripType typ of
  TypeOptional t -> pure t
  _ -> unexpected "optional type" $ show typ

-- | Extract a set of values from a term, mapping a function over each element
-- Example: set string term
set :: Ord x => (Term -> Flow Graph x) -> Term -> Flow Graph (S.Set x)
set f term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermSet s -> S.fromList <$> CM.mapM f (S.toList s)
    _ -> unexpected "set" $ show term

-- | Extract the element type from a set type
-- Example: setType typ
setType :: Type -> Flow s Type
setType typ = case stripType typ of
  TypeSet t -> pure t
  _ -> unexpected "set type" $ show typ

-- * Products and sums

-- | Extract a pair of values from a term, applying functions to each component
-- Example: pair string int32 term
pair :: (Term -> Flow Graph k) -> (Term -> Flow Graph v) -> Term -> Flow Graph (k, v)
pair kf vf term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermProduct terms -> case terms of
      [kTerm, vTerm] -> do
        kVal <- kf kTerm
        vVal <- vf vTerm
        return (kVal, vVal)
      _ -> unexpected "pair" $ show term
    _ -> unexpected "product" $ show term

-- | Extract the component types from a product type
-- Example: productType typ
productType :: Type -> Flow s [Type]
productType typ = case stripType typ of
  TypeProduct types -> pure types
  _ -> unexpected "product type" $ show typ

-- | Extract the component types from a sum type
-- Example: sumType typ
sumType :: Type -> Flow s [Type]
sumType typ = case stripType typ of
  TypeSum types -> pure types
  _ -> unexpected "sum type" $ show typ

-- * Function types

-- | Extract a function type from a type
-- Example: functionType typ
functionType :: Type -> Flow s FunctionType
functionType typ = case stripType typ of
  TypeFunction ft -> pure ft
  _ -> unexpected "function type" $ show typ

-- | Extract a lambda from a term
-- Example: lambda term
lambda :: Term -> Flow Graph Lambda
lambda term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermFunction (FunctionLambda l) -> pure l
    _ -> unexpected "lambda" $ show term

-- | Extract the body of a lambda term
-- Example: lambdaBody term
lambdaBody :: Term -> Flow Graph Term
lambdaBody term = Hydra.Core.lambdaBody <$> lambda term

-- * Let expressions

-- | Extract a let expression from a term
-- Example: letTerm term
letTerm :: Term -> Flow Graph Let
letTerm term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermLet lt -> pure lt
    _ -> unexpected "let term" $ show term

-- | Extract a binding with the given name from a let term
-- Example: letBinding "x" term
letBinding :: String -> Term -> Flow Graph Term
letBinding n term = do
  bindings <- letBindings <$> letTerm term
  case L.filter (\b -> letBindingName b == Name n) bindings of
    [] -> fail $ "no such binding: " ++ n
    [b] -> pure $ letBindingTerm b
    _ -> fail $ "multiple bindings named " ++ n

-- * Records, unions, and newtypes

-- | Extract case statement from a term
-- Example: cases (Name "Result") term
cases :: Name -> Term -> Flow Graph CaseStatement
cases name term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermFunction (FunctionElimination (EliminationUnion cs)) -> if caseStatementTypeName cs == name
      then pure cs
      else unexpected ("case statement for type " ++ unName name) $ show term
    _ -> unexpected "case statement" $ show term

-- | Extract a specific case handler from a case statement term
-- Example: caseField (Name "Result") "success" term
caseField :: Name -> String -> Term -> Flow Graph Field
caseField name n term = do
  cs <- cases name term
  let matching = L.filter (\f -> fieldName f == Name n) $ caseStatementCases cs
  if L.null matching
    then fail $ "not enough cases"
    else pure $ L.head matching

-- | Extract a field value from a list of fields
-- Example: field (Name "name") string fields
field :: Name -> (Term -> Flow Graph x) -> [Field] -> Flow Graph x
field fname mapping fields = case L.filter (\f -> fieldName f == fname) fields of
  [] -> fail $ "field " ++ unName fname ++ " not found"
  [f] -> (stripAndDereferenceTerm $ fieldTerm f) >>= mapping
  _ -> fail $ "multiple fields named " ++ unName fname

-- | Extract a record's fields from a term
-- Example: record (Name "Person") term
record :: Name -> Term -> Flow Graph [Field]
record expected term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermRecord (Record actual fields) -> if actual == expected
      then pure fields
      else unexpected ("record of type " ++ unName expected) (unName actual)
    _ -> unexpected "record" $ show term

-- | Extract the field types from a record type
-- Example: recordType (Name "Person") typ
recordType :: Name -> Type -> Flow s [FieldType]
recordType ename typ = case stripType typ of
  TypeRecord (RowType tname fields) -> if tname == ename
    then pure fields
    else unexpected ("record of type " ++ unName ename) $ "record of type " ++ unName tname
  _ -> unexpected "record type" $ show typ

-- | Extract a unit value (empty record) from a term
-- Example: unit term
unit :: Term -> Flow Graph ()
unit term0 = do
  term <- stripAndDereferenceTerm term0
  fields <- record _Unit term
  if L.null fields
    then pure ()
    else unexpected "unit" $ show term

-- | Extract a field from a union term
-- Example: injection (Name "Result") term
injection :: Name -> Term -> Flow Graph Field
injection expected term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermUnion (Injection actual field) -> if actual == expected
      then pure field
      else unexpected ("injection of type " ++ unName expected) (unName actual)
    _ -> unexpected "injection" $ show term

-- | Extract a field from a union term (alias for injection)
-- Example: variant (Name "Result") term
variant :: Name -> Term -> Flow Graph Field
variant = injection

-- | Extract a unit variant (a variant with an empty record value) from a union term
-- Example: unitVariant (Name "Option") term
unitVariant :: Name -> Term -> Flow Graph Name
unitVariant tname term = do
  field <- variant tname term
  unit $ fieldTerm field
  pure $ fieldName field

-- | Extract the field types from a union type
-- Example: unionType (Name "Result") typ
unionType :: Name -> Type -> Flow s [FieldType]
unionType ename typ = case stripType typ of
  TypeUnion (RowType tname fields) -> if tname == ename
    then pure fields
    else unexpected ("union of type " ++ unName ename) $ "union of type " ++ unName tname
  _ -> unexpected "union type" $ show typ

-- | Extract the wrapped value from a wrapped term
-- Example: wrap (Name "Email") term
wrap :: Name -> Term -> Flow Graph Term
wrap expected term0 = do
  term <- stripAndDereferenceTerm term0
  case term of
    TermWrap (WrappedTerm actual term) -> if actual == expected
      then pure term
      else unexpected ("wrapper of type " ++ unName expected) (unName actual)
    _ -> unexpected ("wrap(" ++ unName expected ++ ")") $ show term

-- | Extract the wrapped type from a wrapper type
-- Example: wrappedType (Name "Email") typ
wrappedType :: Name -> Type -> Flow s Type
wrappedType ename typ = case stripType typ of
  TypeWrap (WrappedType tname t) -> if tname == ename
    then pure t
    else unexpected ("wrapped type " ++ unName ename) $ "wrapped type " ++ unName tname
  _ -> unexpected "wrapped type" $ show typ
